File: myFinance_improved_V2/lib/features/cash_ending/domain/usecases/create_error_adjustment_usecase.dart
import 'package:intl/intl.dart';

import '../../../cash_location/domain/constants/account_ids.dart';
import '../repositories/journal_repository.dart';

/// Parameters for creating an error adjustment journal entry
class CreateErrorAdjustmentParams {
  final double differenceAmount;
  final String companyId;
  final String userId;
  final String locationName;
  final String cashLocationId;
  final String? storeId;

  CreateErrorAdjustmentParams({
    required this.differenceAmount,
    required this.companyId,
    required this.userId,
    required this.locationName,
    required this.cashLocationId,
    this.storeId,
  });
}

/// Use case for creating error adjustment journal entry
/// This adjusts cash discrepancies by creating a journal entry
/// to match the book balance with the actual counted amount
class CreateErrorAdjustmentUseCase {
  final JournalRepository repository;

  CreateErrorAdjustmentUseCase(this.repository);

  Future<Map<String, dynamic>> call(CreateErrorAdjustmentParams params) async {
    // Use centralized account IDs
    const cashAccountId = AccountIds.cash;
    const errorAccountId = AccountIds.errorAdjustment;

    // Get current date
    final now = DateTime.now().toLocal();
    final entryDate = DateFormat('yyyy-MM-ddTHH:mm:ss').format(now);

    // Calculate absolute amount
    final absAmount = params.differenceAmount.abs();
    final isPositiveDifference = params.differenceAmount > 0;

    // Create journal lines
    final lines = [
      {
        'account_id': cashAccountId,
        'description': 'Make error',
        'debit': isPositiveDifference ? absAmount : 0,
        'credit': isPositiveDifference ? 0 : absAmount,
        'cash': {
          'cash_location_id': params.cashLocationId,
        },
      },
      {
        'account_id': errorAccountId,
        'description': 'Make error',
        'debit': isPositiveDifference ? 0 : absAmount,
        'credit': isPositiveDifference ? absAmount : 0,
      },
    ];

    return repository.insertJournalWithEverything(
      baseAmount: absAmount,
      companyId: params.companyId,
      createdBy: params.userId,
      description: 'Make Error - ${params.locationName}',
      entryDate: entryDate,
      lines: lines,
      counterpartyId: null,
      ifCashLocationId: null,
      storeId: params.storeId,
    );
  }
}

File: website/src/features/cash_ending/domain/usecases/CreateErrorAdjustmentUseCase.ts
/**
 * CreateErrorAdjustmentUseCase
 * Business logic for creating error adjustment journal entries
 */

import type { IJournalRepository, JournalLine } from '../repositories/IJournalRepository';

export interface CreateErrorAdjustmentParams {
  differenceAmount: number;
  companyId: string;
  userId: string;
  locationName: string;
  cashLocationId: string;
  storeId: string;
}

export class CreateErrorAdjustmentUseCase {
  constructor(private readonly repository: IJournalRepository) {}

  async execute(params: CreateErrorAdjustmentParams): Promise<{ success: boolean; error?: string }> {
    const { differenceAmount, companyId, userId, locationName, cashLocationId, storeId } = params;

    // Validation
    if (Math.abs(differenceAmount) < 0.01) {
      return { success: false, error: 'No difference to adjust' };
    }

    // Calculate amounts
    const absAmount = Math.abs(differenceAmount);
    const isPositiveDifference = differenceAmount > 0;

    // Create journal lines
    const lines: JournalLine[] = [
      {
        account_id: this.repository.accountIds.cash,
        description: 'Make error',
        debit: isPositiveDifference ? absAmount : 0,
        credit: isPositiveDifference ? 0 : absAmount,
        cash: { cash_location_id: cashLocationId },
      },
      {
        account_id: this.repository.accountIds.errorAdjustment,
        description: 'Make error',
        debit: isPositiveDifference ? 0 : absAmount,
        credit: isPositiveDifference ? absAmount : 0,
      },
    ];

    // Get current date in UTC format
    const now = new Date();
    const entryDateUtc = now.toISOString().replace('T', ' ').substring(0, 19);

    return this.repository.insertJournalWithEverything({
      baseAmount: absAmount,
      companyId,
      createdBy: userId,
      description: `Make Error - ${locationName}`,
      entryDateUtc,
      lines,
      storeId,
    });
  }
}
